ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,169 @@
+    /*
+     * Check permissions for CREATE STREAM and CREATE TABLE statements that have
+     * a persistent query in it.
+     *
+     * This only validates the user can CREATE and WRITE the target topic. The persistent query
+     * should be validated by verifyQuerySourcesPermissions() when the logical plan is
+     * created (see EngineExecutor.execute)
+     */
+
+    final String kafkaTopic = createAsSelect.getName().getSuffix();
+    checkAccess(kafkaTopic, Sets.newHashSet(AclOperation.CREATE, AclOperation.WRITE));
+  }
+
+  private void verifyInsertIntoPermissions(final InsertInto insertInto) {
+    /*
+     * Check permissions for INSERT INTO statements that have a persistent query in it.
+     *
+     * This only validates the user can WRITE the target topic. The persistent query
+     * should be validated by verifyQuerySourcesPermissions() when the logical plan is
+     * created (see EngineExecutor.execute)
+     */
+
+    final String kafkaTopic = insertInto.getTarget().getSuffix();
+    checkAccess(kafkaTopic, Collections.singleton(AclOperation.WRITE));
+  }
+
+  private void verifyInsertValuesPermissions(final InsertValues insertValues) {
+    /*
+     * Check permissions for INSERT INTO VALUES statements that do not have
+     * a persistent query in it.
+     */
+
+    final String kafkaTopic = insertValues.getTarget().getSuffix();
+    checkAccess(kafkaTopic, Collections.singleton(AclOperation.WRITE));
+  }
+
+  private void verifyCreateSourcePermissions(final CreateSource createSource) {
+    /*
+     * Check permissions for CREATE STREAM and CREATE TABLE statements that do not have
+     * a persistent query in it. These statements come the WITH kafka_topic property.
+     */
+    final String kafkaTopic = createSource.getProperties().getKafkaTopic();
+    checkAccess(kafkaTopic, Collections.singleton(AclOperation.READ));
+  }
+
+  /**
+   * Walks through the PlanNode to find all the data sources of the query node.
+   */
+  private void collectKafkaTopics(final Set<String> kafkaTopics, final PlanNode node) {
+    if (node == null) {
+      return;
+    }
+
+    if (node instanceof DataSourceNode) {
+      kafkaTopics.add(((DataSourceNode) node).getDataSource().getKafkaTopicName());
+      return;
+    }
+
+    for (PlanNode nodeSource : node.getSources()) {
+      collectKafkaTopics(kafkaTopics, nodeSource);
+    }
+  }
+
+  /**
+   * Checks if the ServiceContext has access to the topic with the specified AclOperation list.
+   */
+  private void checkAccess(final String topicName, final Set<AclOperation> operations) {
+    try {
+      final TopicDescription topicDescription =
+          serviceContext.getTopicClient().describeTopic(topicName);
+
+      if (!operations.stream().allMatch(topicDescription.authorizedOperations()::contains)) {
+        throw new KsqlTopicAccessException(topicName);
+      }
+    } catch (final Exception e) {
+      throw new KsqlTopicAccessException(topicName, e);
```

Comment:
we should have a more descriptive error message here and above: 
* `"Do not have %s access for topic %s".format(operations, topicName)`
* `"Could not obtain permissions from Kafka for topic: " + topicName`

Reply:
I tried to keep the same message I saw on the `StreamedQueryResource`. However, I made some other changes in the class, so I ended up using your suggestion.